### Exercise: Employee Performance Analysis

**Scenario:**

You work as a human resources analyst for a company that conducts performance evaluations across several departments, including Sales, Marketing, and Operations. Each evaluation entry includes an employee ID, the department, and a performance rating from 1 to 5 (where 5 indicates the highest level of performance). Your task is to analyze these evaluations to identify areas of excellence and areas needing improvement.

**Using this SQL script: https://drive.google.com/file/d/1AwO3Dm7vf72x4ZDZJaEV0TEbiOrxR2YS/view?usp=sharing**

**Objective:**

- Clean up the data's duplicates.
- Remove rows with blank fields.
- Calculate the average salary for each department.
- Identify the department with the highest and lowest salary.
- Suggest actionable insights based on the analysis.

In [ ]:
%load_ext sql

In [80]:
%sql sqlite:///pandas-exercise.db

In [81]:
%%sql
-- Create table
CREATE TABLE Employee (
    employee_id INTEGER,
    first_name VARCHAR(50),
    last_name VARCHAR(50),
    department VARCHAR(50),
    salary DECIMAL(10, 2)
);

-- Insert 100 entries with duplicates and empty fields
INSERT INTO Employee (employee_id, first_name, last_name, department, salary)
VALUES
(1, 'John', 'Doe', 'Sales', 50000.00),
(2, 'Jane', 'Smith', 'Marketing', 60000.00),
(3, 'Michael', 'Johnson', 'Operations', 55000.00),
(4, 'Emily', 'Brown', 'Human Resources', 58000.00),
(5, 'William', 'Jones', 'Sales', 52000.00),
(6, 'Olivia', 'Taylor', 'Marketing', 62000.00),
(7, 'James', 'Williams', 'Operations', 56000.00),
(8, 'Emma', 'Anderson', 'Human Resources', 59000.00),
(9, 'Alexander', 'Martinez', 'Sales', 53000.00),
(10, 'Sophia', 'Garcia', 'Marketing', 63000.00),
-- Duplicate entry
(10, 'Sophia', 'Garcia', 'Marketing', 63000.00),
(11, 'Benjamin', 'Hernandez', 'Operations', 57000.00),
(12, 'Ava', 'Lopez', 'Human Resources', 60000.00),
-- Entry with empty department
(13, 'Ethan', 'Gonzalez', '', 54000.00),
(14, 'Mia', 'Wilson', 'Marketing', 64000.00),
(15, 'Daniel', 'Perez', 'Operations', 59000.00),
-- Entry with empty salary
(16, 'Charlotte', 'Rivera', 'Human Resources', NULL),
(17, 'Logan', 'Tomas', 'Sales', 55000.00),
(18, 'Harper', 'Moore', 'Marketing', 65000.00),
(19, 'Matthew', 'Wright', 'Operations', 60000.00),
-- Duplicate entry
(19, 'Matthew', 'Wright', 'Operations', 60000.00),
(20, 'Evelyn', 'Nguyen', 'Human Resources', 62000.00),
(21, 'Ryan', 'Young', 'Sales', 56000.00),
(22, 'Amelia', 'King', 'Marketing', 66000.00),
(23, 'Noah', 'Scott', 'Operations', 61000.00),
(24, 'Liam', 'Gomez', 'Human Resources', 63000.00),
(25, 'Isabella', 'Cook', 'Sales', 57000.00),
-- Entry with empty last name
(26, 'Aiden', '', 'Marketing', 67000.00),
(27, 'Ella', 'Green', 'Operations', 62000.00),
(28, 'Lucas', 'Adams', 'Human Resources', 64000.00),
(29, 'Chloe', 'Baker', 'Sales', 58000.00),
(30, 'Avery', 'Nelson', 'Marketing', 68000.00),
-- Entry with empty first name
(31, '', 'Carter', 'Operations', 63000.00),
(32, 'Luna', 'Evans', 'Human Resources', 65000.00),
(33, 'Elijah', 'Parker', 'Sales', 59000.00),
(34, 'Layla', 'Mitchell', 'Marketing', 69000.00),
(35, 'Carter', 'Roberts', 'Operations', 64000.00),
-- Duplicate entry
(35, 'Carter', 'Roberts', 'Operations', 64000.00),
(36, 'Zoey', 'James', 'Human Resources', 66000.00),
(37, 'Alexander', 'Gray', 'Sales', 60000.00),
(38, 'Penelope', 'Hill', 'Marketing', 70000.00),
(39, 'Sebastian', 'Adams', 'Operations', 65000.00),
(40, 'Mila', 'Campbell', 'Human Resources', 67000.00),
-- Entry with empty department and salary
(41, 'Michael', 'Jenkins', '', NULL),
(42, 'Riley', 'Perez', 'Marketing', 71000.00),
(43, 'Levi', 'Robinson', 'Operations', 66000.00),
(44, 'Zoe', 'Phillips', 'Human Resources', 68000.00),
(45, 'Nathan', 'Reed', 'Sales', 62000.00),
(46, 'Nora', 'Torres', 'Marketing', 72000.00),
(47, 'David', 'Turner', 'Operations', 67000.00),
(48, 'Sofia', 'Parker', 'Human Resources', 69000.00),
-- Entry with empty last name and salary
(49, 'Isaac', '', 'Sales', NULL),
(50, 'Ellie', 'Edwards', 'Marketing', 73000.00),
(51, 'Henry', 'Stewart', 'Operations', 68000.00),
(52, 'Stella', 'Collins', 'Human Resources', 70000.00),
(53, 'Gabriel', 'Harris', 'Sales', 64000.00),
(54, 'Hannah', 'Stewart', 'Marketing', 74000.00),
(55, 'Josiah', 'Martin', 'Operations', 69000.00),
(56, 'Madison', 'Morris', 'Human Resources', 71000.00),
(57, 'Luke', 'Lee', 'Sales', 65000.00),
(58, 'Aria', 'Hall', 'Marketing', 75000.00),
(59, 'Caleb', 'Thompson', 'Operations', 70000.00),
(60, 'Grace', 'Gonzalez', 'Human Resources', 72000.00),
(61, 'Owen', 'Clark', 'Sales', 66000.00),
(62, 'Scarlett', 'Moore', 'Marketing', 76000.00),
-- Duplicate entry
(62, 'Scarlett', 'Moore', 'Marketing', 76000.00),
(63, 'Wyatt', 'Allen', 'Operations', 71000.00),
(64, 'Aubrey', 'Harris', 'Human Resources', 73000.00),
(65, 'Isaiah', 'Lewis', 'Sales', 67000.00),
(66, 'Adeline', 'King', 'Marketing', 77000.00),
(67, 'Henry', 'King', 'Operations', 72000.00),
(68, 'Leah', 'Nelson', 'Human Resources', 74000.00),
(69, 'Cameron', 'Brown', 'Sales', 68000.00),
(70, 'Hazel', 'Hall', 'Marketing', 78000.00),
-- Entry with empty first name and department
(71, '', 'Young', '', 73000.00),
(72, 'Zoey', 'Hernandez', 'Human Resources', 75000.00),
(73, 'Dylan', 'Hall', 'Sales', 69000.00),
(74, 'Violet', 'Scott', 'Marketing', 79000.00),
(75, 'Tyler', 'Adams', 'Operations', 74000.00),
(76, 'Claire', 'King', 'Human Resources', 76000.00),
(77, 'Julian', 'Hernandez', 'Sales', 70000.00),
(78, 'Lily', 'Young', 'Marketing', 80000.00),
(79, 'Eli', 'Martinez', 'Operations', 75000.00),
(80, 'Brooklyn', 'Hill', 'Human Resources', 77000.00),
(81, 'Nolan', 'Harris', 'Sales', 71000.00),
-- Entry with empty last name and department
(82, 'Anna', '', '', 81000.00),
(83, 'Luke', 'Johnson', 'Operations', 76000.00),
(84, 'Elena', 'Hernandez', 'Human Resources', 78000.00),
(85, 'Aaron', 'Martinez', 'Sales', 72000.00),
(86, 'Audrey', 'Martinez', 'Marketing', 82000.00),
(87, 'Eliana', 'Gomez', 'Operations', 77000.00),
(88, 'Ezra', 'Perez', 'Human Resources', 79000.00),
(89, 'Evan', 'Wilson', 'Sales', 73000.00),
(90, 'Gabriella', 'Rodriguez', 'Marketing', 83000.00),
(91, 'Xavier', 'Taylor', 'Operations', 78000.00),
(92, 'Lucy', 'Ramirez', 'Human Resources', 80000.00),
(93, 'Thomas', 'Torres', 'Sales', 74000.00),
(94, 'Eleanor', 'Anderson', 'Marketing', 84000.00),
(95, 'Leonardo', 'Hill', 'Operations', 79000.00),
(96, 'Mackenzie', 'Lopez', 'Human Resources', 81000.00),
(97, 'Carson', 'Flores', 'Sales', 75000.00),
(98, 'Naomi', 'Reed', 'Marketing', 85000.00),
(99, 'Jaxon', 'Gonzalez', 'Operations', 80000.00),
(100, 'Paisley', 'Hernandez', 'Human Resources', 82000.00),
-- Entry with empty salary
(101, 'Test', 'Employee', 'Sales', NULL);

 * sqlite:///pandas-exercise.db
(sqlite3.OperationalError) table Employee already exists
[SQL: -- Create table
CREATE TABLE Employee (
    employee_id INTEGER,
    first_name VARCHAR(50),
    last_name VARCHAR(50),
    department VARCHAR(50),
    salary DECIMAL(10, 2)
);]
(Background on this error at: https://sqlalche.me/e/20/e3q8)


In [82]:
import sqlite3
import pandas as pd

conn = sqlite3.connect('pandas-exercise.db')
df = pd.read_sql_query('SELECT * FROM Employee', conn)

# Remove Duplicates
df.drop_duplicates(inplace=True)

# Drop NA Values
df.replace('', pd.NA, inplace=True)
df.dropna(inplace=True)

df

employee_id first_name  last_name       department   salary
0              1       John        Doe            Sales  50000.0
1              2       Jane      Smith        Marketing  60000.0
2              3    Michael    Johnson       Operations  55000.0
3              4      Emily      Brown  Human Resources  58000.0
4              5    William      Jones            Sales  52000.0
..           ...        ...        ...              ...      ...
99            96  Mackenzie      Lopez  Human Resources  81000.0
100           97     Carson     Flores            Sales  75000.0
101           98      Naomi       Reed        Marketing  85000.0
102           99      Jaxon   Gonzalez       Operations  80000.0
103          100    Paisley  Hernandez  Human Resources  82000.0

[92 rows x 5 columns]

In [83]:
# Calculate average salary per department
avg_salary = df.groupby('department')['salary'].mean()

avg_salary

department
Human Resources    70375.000000
Marketing          72869.565217
Operations         67869.565217
Sales              63454.545455
Name: salary, dtype: float64

In [84]:
# Highest and Lowest Salary Departments

# .idxmax() gets the index of the highest, idxmin() lowest
highest = avg_salary.idxmax()
lowest = avg_salary.idxmin()

print(f'Highest: {highest} | Lowest: {lowest}')

Highest: Marketing | Lowest: Sales


# **Report**

Here are the results for the average salaries per department:

| Department | Average Salary 
|------------|----------------
| Human Resources | 70375.000000
| Marketing | 72869.565217
| Operations | 67869.565217
| Sales | 63454.545455

## **Results**

- Highest Budget Department: Marketing ($72,869.57)
- Lowest Budget Department: Sales ($63,454.55)

## Actionable Insights:

- Marketing has the highest budget. Investigate if this aligns with the company's goals and priorities.
- Sales has the lowest budget. Analyze if this budget is sufficient to achieve sales targets. 
- Consider reallocating budget if necessary.